akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -153,6 +172,7 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
     newState match {
       case Closed =>
         streamStates -= streamId
+        if (streamStates.isEmpty && completing) onComplete()
```

Comment:
since `completing` is only set when `substreamIn` completes, I wonder if it would be simpler to check if `isClosed(substreamIn)` here

Reply:
I like the idea but I’ve tried and can’t seem to make it work. 

Either (1) I overwrite the `onUpstreamFinish` for `substreamIn` to do nothing (then it’s not closed), or (2) implement `onUpstreamFinish` to `cancel(substreamIn)` which seems to not close the stage (?).

For the record, code looked like:

```scala
      override def complete(): Unit = {
        if (!isClosed(substreamIn)) // may be close if the invocation to complete() was delayed
          cancel(substreamIn)
        cancel(frameIn)
        complete(frameOut)
        bufferedSubStreamOutput.complete()
      }
      override def completing(): Boolean = isClosed(substreamOut)
      setHandler(substreamIn, new InHandler {
	    ...
        override def onUpstreamFinish(): Unit = {
          if (activeStreamCount() == 0) complete() // no buffer so proceed with complete()
          else cancel(substreamIn)  // work pending, close the inlet

          // FIXME: start a timer
          //  - on timer, log the timer event and proceed with the finish logic
        }
      })
```

I think it's good enough as we have it.